metric_providers/powermetrics/provider.py

```diff
@@ -33,13 +35,35 @@ def __init__(self, resolution):
             '-o',
             self._filename]
 
+    def is_powermetrics_running(self):
+        try:
+            output = subprocess.check_output('pgrep -x powermetrics', shell=True)
+            if output.strip():  # If the output is not empty, the process is running.
+                return True
+        except subprocess.CalledProcessError:  # If the process is not running, 'pgrep' returns non-zero exit code.
+            return False
+
+
     def stop_profiling(self):
         try:
             # We try calling the parent method but if this doesn't work we use the more hardcore approach
             super().stop_profiling()
         except PermissionError:
-            # This isn't the nicest way of doing this but there isn't really any other way that is nicer
+            #This isn't the nicest way of doing this but there isn't really any other way that is nicer
             subprocess.check_output('sudo /usr/bin/killall powermetrics', shell=True)
+            print('Killed powermetrics process with killall!')
+
+        # As killall returns right after sending the SIGKILL we need to wait and make sure that the process
+        # had time to flush everything to disk
+        count = 0
+        while self.is_powermetrics_running():
+            time.sleep(1)
+            count += 1
+            if count >= 5:
+                raise RuntimeError('powermetrics is not stopping. Please kill with `kill -9 ...`.')
```

Comment:
If it needs to be anyway killed with kill -9, why don't we do it?

This would then not only leave the output file corrupted, but also the state of the GMT, as you might have two instances of powermetrics running. correct?

Reply:
Has been stale for a week now. If you do not have any further input I would vouch for changing the code to be less manual and more automatic.

Reply:
Totally. Is on my todo.

Reply:
I now remember why I did it like this. `kill -9` needs to be called as root and I was hesitant to add another line to sudoers. No real reason why to be honest. Added it.